Allow Externally Managed CRD Fields

dalton-hill-0 opened the pull request:
### Description of your changes
Use server-side apply for reconciling CRDs for composite resources and claims. This will allow for other processes (e.g., CertManager) to control fields on the CRD that are not owned by the reconciler.

*Note: The current implementation only fixes composite resource CRDs. If this approach is desired, I will implement for Claims as well.*

Fixes #5723 

I have:

- [X] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] Added or updated unit tests.
- [ ] ~Added or updated e2e tests.~
- [ ] ~Linked a PR or a [docs tracking issue] to [document this change].~
- [ ] ~Added `backport release-x.y` labels to auto-backport this PR.~

negz commented on internal/controller/apiextensions/definition/reconciler.go:
```diff
@@ -267,8 +282,9 @@ func NewReconciler(mgr manager.Manager, opts ...ReconcilerOption) *Reconciler {
 type Reconciler struct {
-	client resource.ClientApplicator
-	mgr    manager.Manager
+	client    resource.ClientApplicator
```
Does this still need to be a ClientApplicator? We don't call the Apply method anymore, right?

negz commented on internal/controller/apiextensions/definition/reconciler.go:
```diff
@@ -209,6 +216,14 @@ func WithClientApplicator(ca resource.ClientApplicator) ReconcilerOption {
 	}
 }
 
+// WithClientset specifies how the Reconciler should interact with the
+// Kubernetes API.
```
It always bothers me a bit when we have to use both a controller-runtime client and a client-go clientset. I'm guessing its unavoidable in this case right, because the controller-runtime client can't SSA typed objects (only `*Unstructured`)?

negz commented on internal/controller/apiextensions/definition/reconciler.go:
```diff
@@ -278,13 +294,16 @@ type Reconciler struct {
 
 	options apiextensionscontroller.Options
+
+	name string
```
Is this used anywhere?

negz commented on internal/xcrd/crd.go:
```diff
@@ -97,6 +101,117 @@ func ForCompositeResource(xrd *v1.CompositeResourceDefinition) (*extv1.CustomRes
 	return crd, nil
 }
 
+func ModifyApplyConfigurationForCompositeResource(crd *appcfgextv1.CustomResourceDefinitionApplyConfiguration, xrd *v1.CompositeResourceDefinition) error { //nolint:gocognit
```
Needing to do all of this feels a little like it defeats the point of SSA (i.e. just form a simple intent and send it). I wonder if it would be simpler to just do a get-mutate-update of the CRD? I think we could assume exclusive ownership of the CRD spec (i.e. just overwrite that each time).

dalton-hill-0 commented:
@negz 
I took an entirely different approach (no longer using `applyconfigurations` libraries) to address your comments. This just removes the applicator and replaces `Apply` with `Patch`. I grabbed the patch args from `syncer_ssa.go`.
I tested this and it still allows CertManager to function as desired.

negz commented:
@dalton-hill-0 This looks great, and is definitely ideally how SSA would work.

Unfortunately I think there are bugs when you SSA a 'real' type (i.e. a CRD) using the controller-runtime client today. See https://github.com/kubernetes-sigs/controller-runtime/issues/347#issuecomment-1470092104.

dalton-hill-0 commented:
> Unfortunately I think there are bugs when you SSA a 'real' type (i.e. a CRD) using the controller-runtime client today.

That is indeed unfortunate.

> Needing to do all of this feels a little like it defeats the point of SSA (i.e. just form a simple intent and send it). I wonder if it would be simpler to just do a get-mutate-update of the CRD? I think we could assume exclusive ownership of the CRD spec (i.e. just overwrite that each time).

I agree, and it introduces a lot of new code that could introduce bugs. However, CertManager needs to own the field at `spec.conversion.webhook.clientConfig.caBundle` so overriding `spec` each time would prevent the desired behavior of this PR (unless if I'm thinking of this wrong?).

A couple of ideas:
- Maybe we can just use `Patch` anyways? It has some bugs, but do these bugs affect how Crossplane handles CRDs? I assume we are already affected by zero-value defaults with how we handle CRDs currently. Are there other bugs I'm not aware of that would also affect CRDs?
- We could use one of the other approaches mentioned in the [issue](https://github.com/crossplane/crossplane/issues/5723) (see below)

>Hard-Coded Approach
Add hard-coded logic in Crossplane's Claim CRD and CompositeResource CRD reconcilers that is specific to CertManager. This logic would check for a CA Bundle in the existing CRD and copy it over before applying the spec derived from the XRD.

>Generic Approach
Extend the XRD API to allow users to specify paths that are externally managed (e.g., by CertManger). For each path, the CRD reconciler would attempt to copy the value from existing CRD before applying the CRD derived from the XRD.

negz commented:
> Maybe we can just use Patch anyways? It has some bugs, but do these bugs affect how Crossplane handles CRDs? I assume we are already affected by zero-value defaults with how we handle CRDs currently. Are there other bugs I'm not aware of that would also affect CRDs?

This sounds good to me - if it looks like we're not actually affected in practice I'm happy to stick with the implementation you have here and maybe add a comment mentioning that we're aware of the issue but don't seem to be affected.

dalton-hill-0 commented:
I added comments and updated tests.
Additionally I checked some diffs between the CRDs that get generated on the main branch vs this branch to make sure nothing changed.

negz commented:
@dalton-hill-0 I noticed you closed this - did you find that this approach wouldn't work?

dalton-hill-0 commented:
> @dalton-hill-0 I noticed you closed this - did you find that this approach wouldn't work?

No that was on accident. I accidentally synced the wrong fork branch and I think it closed this as a result.

negz commented on internal/controller/apiextensions/definition/reconciler.go:
```diff
@@ -427,8 +431,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req reconcile.Request) (reco
 
-	origRV := ""
-	if err := r.client.Apply(ctx, crd, resource.MustBeControllableBy(d.GetUID()), resource.StoreCurrentRV(&origRV)); err != nil {
+	// We are aware that using controller-runtime server-side apply will result in
+	// some zero-value fields, however, this should not affect the way CRDs are
+	// implemented in this controller. For a discussion on the runtime issue, see
+	// https://github.com/kubernetes-sigs/controller-runtime/issues/347.
```
@turkenh Could you do a quick sanity check on this. I know you went pretty deep on SSA issues recently for provider-kubernetes.

I feel okay switching to SSA here as long as we don't see any issues applying CRDs in practice (which appears to be the case in @dalton-hill-0's testing.)

dalton-hill-0 replied:
Just a friendly reminder 🙂
@turkenh

negz replied:
@dalton-hill-0 It looks like @turkenh is on vacation. He'll be back in a week.

turkenh replied:
This sounds good to me. 

To see whether SSA causing any immediate problem, a quick sanity test could be leaving a watch request for a CRD of an XRD open for a while, e.g. `kubectl get crds <crd-name> -w`, and checking whether it gets periodic updates or not (i.e. resourceVersion keeps increasing). You may restart Crossplane or add some labels to XRDs to trigger new reconciliations.

dalton-hill-0 replied:
I performed the sanity checks above and all behavior was expected.
- no increase to resource version on crossplane restart
- no increase to resource version when adding annotations to xrds (they do not get copied over from meta)
- resource version increased when adding label to xrd (copies over to crds)
- no infinite patch loop between crossplane and cert manager

dalton-hill-0 commented:
Rebased from master and re-tested.

dalton-hill-0 commented:
From what I can tell, the e2e tests that are failing are expected with the change and don't look concerning. If you agree, I'll update the tests to pass.

Avarei commented:
It's sad to see that this PR is almost done, just to be auto-closed. Can I help in any way to get it back on track?

dalton-hill-0 commented:
> It's sad to see that this PR is almost done, just to be auto-closed. Can I help in any way to get it back on track?

I think the only thing that needs to be done is for the team to weigh the potential risks of the behavior change.

zonybob commented:
Our team would also love to implement conversion webhooks, and I believe we are currently blocked on this behavior (CRD caBundle contention between cert-manager and Crossplane). Is there anything we could do revive the conversation around this pull request?
